### DISCUSSION TITLE
TypeError: numpy boolean subtract, the `-` operator, is not supported, use the bitwise_xor, the `^` operator, or the logical_xor function instead.

### DISCUSSION THREAD

**Discussion by nril0001** [pybamm-team/PyBaMM#3092] [Q&A]
I am in the process of attempting to install pybamm into a new environment. However, I keep coming across the following error no matter what installation method I have attempted. I know for a fact that the issue does not arise from the file I have been as running as to runs just fine in a different pybamm environment that I have available. I have included the entire error trace back and would appreciate any advice that you could provide on ways that I could potentially fix this issue. 

This is the list of packages that I currently have included in my environment:

name: pybamm
channels:
  - defaults
dependencies:
  - alabaster=0.7.12=pyhd3eb1b0_0
  - anyio=3.5.0=py310haa95532_0
  - argon2-cffi=21.3.0=pyhd3eb1b0_0
  - argon2-cffi-bindings=21.2.0=py310h2bbff1b_0
  - arrow=1.2.3=py310haa95532_1
  - astroid=2.14.2=py310haa95532_0
  - asttokens=2.0.5=pyhd3eb1b0_0
  - atomicwrites=1.4.0=py_0
  - attrs=22.1.0=py310haa95532_0
  - autopep8=1.6.0=pyhd3eb1b0_1
  - babel=2.11.0=py310haa95532_0
  - backcall=0.2.0=pyhd3eb1b0_0
  - bcrypt=3.2.0=py310h2bbff1b_1
  - beautifulsoup4=4.12.2=py310haa95532_0
  - binaryornot=0.4.4=pyhd3eb1b0_1
  - black=23.3.0=py310haa95532_0
  - bleach=4.1.0=pyhd3eb1b0_0
  - brotlipy=0.7.0=py310h2bbff1b_1002
  - bzip2=1.0.8=he774522_0
  - ca-certificates=2023.05.30=haa95532_0
  - certifi=2023.5.7=py310haa95532_0
  - cffi=1.15.1=py310h2bbff1b_3
  - chardet=4.0.0=py310haa95532_1003
  - charset-normalizer=2.0.4=pyhd3eb1b0_0
  - click=8.0.4=py310haa95532_0
  - cloudpickle=2.2.1=py310haa95532_0
  - colorama=0.4.6=py310haa95532_0
  - comm=0.1.2=py310haa95532_0
  - cookiecutter=1.7.3=pyhd3eb1b0_0
  - cryptography=39.0.1=py310h21b164f_0
  - debugpy=1.5.1=py310hd77b12b_0
  - decorator=5.1.1=pyhd3eb1b0_0
  - defusedxml=0.7.1=pyhd3eb1b0_0
  - diff-match-patch=20200713=pyhd3eb1b0_0
  - dill=0.3.6=py310haa95532_0
  - docstring-to-markdown=0.11=py310haa95532_0
  - docutils=0.18.1=py310haa95532_3
  - entrypoints=0.4=py310haa95532_0
  - executing=0.8.3=pyhd3eb1b0_0
  - flake8=6.0.0=py310haa95532_0
  - giflib=5.2.1=h8cc25b3_3
  - glib=2.69.1=h5dc1a3c_2
  - gst-plugins-base=1.18.5=h9e645db_0
  - gstreamer=1.18.5=hd78058f_0
  - icu=58.2=ha925a31_3
  - idna=3.4=py310haa95532_0
  - imagesize=1.4.1=py310haa95532_0
  - importlib_metadata=6.0.0=hd3eb1b0_0
  - inflection=0.5.1=py310haa95532_0
  - intervaltree=3.1.0=pyhd3eb1b0_0
  - ipykernel=6.19.2=py310h9909e9c_0
  - ipython=8.12.0=py310haa95532_0
  - ipython_genutils=0.2.0=pyhd3eb1b0_1
  - isort=5.9.3=pyhd3eb1b0_0
  - jaraco.classes=3.2.1=pyhd3eb1b0_0
  - jedi=0.18.1=py310haa95532_1
  - jellyfish=0.9.0=py310h2bbff1b_0
  - jinja2=3.1.2=py310haa95532_0
  - jinja2-time=0.2.0=pyhd3eb1b0_3
  - jpeg=9e=h2bbff1b_1
  - jsonschema=4.17.3=py310haa95532_0
  - jupyter_client=8.1.0=py310haa95532_0
  - jupyter_core=5.3.0=py310haa95532_0
  - jupyter_events=0.6.3=py310haa95532_0
  - jupyter_server=2.5.0=py310haa95532_0
  - jupyter_server_terminals=0.4.4=py310haa95532_1
  - jupyterlab_pygments=0.1.2=py_0
  - keyring=23.13.1=py310haa95532_0
  - krb5=1.19.4=h5b6d351_0
  - lazy-object-proxy=1.6.0=py310h2bbff1b_0
  - lerc=3.0=hd77b12b_0
  - libclang=14.0.6=default_hb5a9fac_1
  - libclang13=14.0.6=default_h8e68704_1
  - libdeflate=1.17=h2bbff1b_0
  - libffi=3.4.4=hd77b12b_0
  - libiconv=1.16=h2bbff1b_2
  - libogg=1.3.5=h2bbff1b_1
  - libpng=1.6.39=h8cc25b3_0
  - libsodium=1.0.18=h62dcd97_0
  - libspatialindex=1.9.3=h6c2663c_0
  - libtiff=4.5.0=h6c2663c_2
  - libvorbis=1.3.7=he774522_0
  - libwebp=1.2.4=hbc33d0d_1
  - libwebp-base=1.2.4=h2bbff1b_1
  - libxml2=2.10.3=h0ad7f3c_0
  - libxslt=1.1.37=h2bbff1b_0
  - lxml=4.9.2=py310h2bbff1b_0
  - lz4-c=1.9.4=h2bbff1b_0
  - markupsafe=2.1.1=py310h2bbff1b_0
  - matplotlib-inline=0.1.6=py310haa95532_0
  - mccabe=0.7.0=pyhd3eb1b0_0
  - mistune=0.8.4=py310h2bbff1b_1000
  - more-itertools=8.12.0=pyhd3eb1b0_0
  - mypy_extensions=0.4.3=py310haa95532_1
  - nbclassic=0.5.5=py310haa95532_0
  - nbclient=0.5.13=py310haa95532_0
  - nbconvert=6.5.4=py310haa95532_0
  - nbformat=5.7.0=py310haa95532_0
  - nest-asyncio=1.5.6=py310haa95532_0
  - notebook=6.5.4=py310haa95532_0
  - notebook-shim=0.2.2=py310haa95532_0
  - numpydoc=1.5.0=py310haa95532_0
  - openssl=1.1.1u=h2bbff1b_0
  - pandocfilters=1.5.0=pyhd3eb1b0_0
  - paramiko=2.8.1=pyhd3eb1b0_0
  - parso=0.8.3=pyhd3eb1b0_0
  - pathspec=0.10.3=py310haa95532_0
  - pcre=8.45=hd77b12b_0
  - pexpect=4.8.0=pyhd3eb1b0_3
  - pickleshare=0.7.5=pyhd3eb1b0_1003
  - pip=23.1.2=py310haa95532_0
  - platformdirs=2.5.2=py310haa95532_0
  - ply=3.11=py310haa95532_0
  - poyo=0.5.0=pyhd3eb1b0_0
  - prometheus_client=0.14.1=py310haa95532_0
  - prompt-toolkit=3.0.36=py310haa95532_0
  - psutil=5.9.0=py310h2bbff1b_0
  - ptyprocess=0.7.0=pyhd3eb1b0_2
  - pure_eval=0.2.2=pyhd3eb1b0_0
  - pycodestyle=2.10.0=py310haa95532_0
  - pycparser=2.21=pyhd3eb1b0_0
  - pydocstyle=6.3.0=py310haa95532_0
  - pyflakes=3.0.1=py310haa95532_0
  - pygments=2.15.1=py310haa95532_1
  - pylint=2.16.2=py310haa95532_0
  - pyls-spyder=0.4.0=pyhd3eb1b0_0
  - pynacl=1.5.0=py310h8cc25b3_0
  - pyopenssl=23.0.0=py310haa95532_0
  - pyqt=5.15.7=py310hd77b12b_0
  - pyqt5-sip=12.11.0=py310hd77b12b_0
  - pyqtwebengine=5.15.7=py310hd77b12b_0
  - pyrsistent=0.18.0=py310h2bbff1b_0
  - pysocks=1.7.1=py310haa95532_0
  - python=3.10.11=h966fe2a_2
  - python-dateutil=2.8.2=pyhd3eb1b0_0
  - python-fastjsonschema=2.16.2=py310haa95532_0
  - python-json-logger=2.0.7=py310haa95532_0
  - python-lsp-black=1.2.1=py310haa95532_0
  - python-lsp-jsonrpc=1.0.0=pyhd3eb1b0_0
  - python-lsp-server=1.7.2=py310haa95532_0
  - python-slugify=5.0.2=pyhd3eb1b0_0
  - pytoolconfig=1.2.5=py310haa95532_1
  - pywin32=305=py310h2bbff1b_0
  - pywin32-ctypes=0.2.0=py310haa95532_1000
  - pywinpty=2.0.10=py310h5da7b33_0
  - pyyaml=6.0=py310h2bbff1b_1
  - pyzmq=25.1.0=py310hd77b12b_0
  - qdarkstyle=3.0.2=pyhd3eb1b0_0
  - qt-main=5.15.2=he8e5bd7_8
  - qt-webengine=5.15.9=hb9a9bb5_5
  - qtconsole=5.4.2=py310haa95532_0
  - qtpy=2.2.0=py310haa95532_0
  - qtwebkit=5.212=h2bbfb41_5
  - requests=2.29.0=py310haa95532_0
  - rfc3339-validator=0.1.4=py310haa95532_0
  - rfc3986-validator=0.1.1=py310haa95532_0
  - rope=1.7.0=py310haa95532_0
  - rtree=1.0.1=py310h2eaa2aa_0
  - send2trash=1.8.0=pyhd3eb1b0_1
  - setuptools=67.8.0=py310haa95532_0
  - sip=6.6.2=py310hd77b12b_0
  - six=1.16.0=pyhd3eb1b0_1
  - sniffio=1.2.0=py310haa95532_1
  - snowballstemmer=2.2.0=pyhd3eb1b0_0
  - sortedcontainers=2.4.0=pyhd3eb1b0_0
  - soupsieve=2.4=py310haa95532_0
  - sphinx=5.0.2=py310haa95532_0
  - sphinxcontrib-applehelp=1.0.2=pyhd3eb1b0_0
  - sphinxcontrib-devhelp=1.0.2=pyhd3eb1b0_0
  - sphinxcontrib-htmlhelp=2.0.0=pyhd3eb1b0_0
  - sphinxcontrib-jsmath=1.0.1=pyhd3eb1b0_0
  - sphinxcontrib-qthelp=1.0.3=pyhd3eb1b0_0
  - sphinxcontrib-serializinghtml=1.1.5=pyhd3eb1b0_0
  - spyder=5.4.3=py310haa95532_1
  - spyder-kernels=2.4.3=py310haa95532_0
  - sqlite=3.41.2=h2bbff1b_0
  - stack_data=0.2.0=pyhd3eb1b0_0
  - terminado=0.17.1=py310haa95532_0
  - text-unidecode=1.3=pyhd3eb1b0_0
  - textdistance=4.2.1=pyhd3eb1b0_0
  - three-merge=0.1.1=pyhd3eb1b0_0
  - tinycss2=1.2.1=py310haa95532_0
  - tk=8.6.12=h2bbff1b_0
  - toml=0.10.2=pyhd3eb1b0_0
  - tomli=2.0.1=py310haa95532_0
  - tomlkit=0.11.1=py310haa95532_0
  - tornado=6.2=py310h2bbff1b_0
  - traitlets=5.7.1=py310haa95532_0
  - typing-extensions=4.6.3=py310haa95532_0
  - typing_extensions=4.6.3=py310haa95532_0
  - ujson=5.4.0=py310hd77b12b_0
  - unidecode=1.2.0=pyhd3eb1b0_0
  - urllib3=1.26.16=py310haa95532_0
  - vc=14.2=h21ff451_1
  - vs2015_runtime=14.27.29016=h5e58377_2
  - watchdog=2.1.6=py310haa95532_0
  - wcwidth=0.2.5=pyhd3eb1b0_0
  - webencodings=0.5.1=py310haa95532_1
  - websocket-client=0.58.0=py310haa95532_4
  - whatthepatch=1.0.2=py310haa95532_0
  - wheel=0.38.4=py310haa95532_0
  - win_inet_pton=1.1.0=py310haa95532_0
  - winpty=0.4.3=4
  - wrapt=1.14.1=py310h2bbff1b_0
  - xz=5.4.2=h8cc25b3_0
  - yaml=0.2.5=he774522_0
  - yapf=0.31.0=pyhd3eb1b0_0
  - zeromq=4.3.4=hd77b12b_0
  - zlib=1.2.13=h8cc25b3_0
  - zstd=1.5.5=hd43e919_0
  - pip:
      - contourpy==1.1.0
      - cycler==0.11.0
      - fonttools==4.40.0
      - future==0.18.3
      - imageio==2.31.1
      - importlib-metadata==6.7.0
      - kiwisolver==1.4.4
      - matplotlib==3.7.1
      - mpmath==1.3.0
      - numpy==1.25.0
      - packaging==23.1
      - pandas==2.0.3
      - pillow==10.0.0
      - pluggy==1.2.0
      - py==1.11.0
      - pybamm==23.5
      - pylint-venv==2.3.0
      - pyparsing==3.1.0
      - pytz==2023.3
      - qstylizer==0.2.2
      - qtawesome==1.2.2
      - scipy==1.11.1
      - sympy==1.12
      - tqdm==4.65.0
      - tzdata==2023.3
      - xarray==2023.6.0
      - zipp==3.15.0

Error traceback

Python 3.10.11 | packaged by Anaconda, Inc. | (main, Apr 20 2023, 18:56:50) [MSC v.1916 64 bit (AMD64)]
Type "copyright", "credits" or "license" for more information.

IPython 8.12.0 -- An enhanced Interactive Python.

runfile('C:/Users/natha/Desktop/Code/main.py', wdir='C:/Users/natha/Desktop/Code')
['DigiElech/2023-06-26 solution/CV/CV_k0_10000_R_800_Cd_1e-9_Ds_2e-5_Dp_1.5e-5.txt', 0.001, 1e-05, 2023.0, -6.0, -26.0, 0.0, 10000.0, 800.0, 1e-09, 2e-05, 1.5e-05]
Traceback (most recent call last):

  File ~\anaconda3\envs\pybamm\lib\site-packages\pybamm\parameters\parameter_values.py:497 in process_symbol
    return self._processed_symbols[symbol]

KeyError: Subtraction(-0x887745c55906d0f, -, children=['Voltage start [V] / (Gas constant [J K-1 mol-1] * Temperature [K] / Faraday Constant [C mol-1]) + (time <= (Voltage start [V] / (Gas constant [J K-1 mol-1] * Temperature [K] / Faraday Constant [C mol-1])) - (Voltage reverse [V] / (Gas constant [J K-1 mol-1] * Temperature [K] / Faraday Constant [C mol-1]))) * -time + ((Voltage start [V] / (Gas constant [J K-1 mol-1] * Temperature [K] / Faraday Constant [C mol-1])) - (Voltage reverse [V] / (Gas constant [J K-1 mol-1] * Temperature [K] / Faraday Constant [C mol-1])) < time) * (time - (2.0 * ((Voltage start [V] / (Gas constant [J K-1 mol-1] * Temperature [K] / Faraday Constant [C mol-1])) - (Voltage reverse [V] / (Gas constant [J K-1 mol-1] * Temperature [K] / Faraday Constant [C mol-1]))))) - (Uncompensated Resistance [Ohm] * Faraday Constant [C mol-1] * Electrode Area [cm2] * Far-field concentration of S(soln) [mol cm-3] * sqrt(Diffusion Coefficient of S [cm2 s-1]) * sqrt(Faraday Constant [C mol-1] * Scan Rate [V s-1] / (Gas constant [J K-1 mol-1] * Temperature [K])) / (Gas constant [J K-1 mol-1] * Temperature [K] / Faraday Constant [C mol-1]) * (Redox Rate [s-1] * sqrt((Gas constant [J K-1 mol-1] * Temperature [K] * Diffusion Coefficient of S [cm2 s-1] / Faraday Constant [C mol-1]) * Scan Rate [V s-1]) / Diffusion Coefficient of S [cm2 s-1] * ((boundary value(P(soln) [non-dim]) * exp((1.0 - Symmetry factor [non-dim]) * (Effective Voltage [non-dim] - (Reversible Potential [V] / (Gas constant [J K-1 mol-1] * Temperature [K] / Faraday Constant [C mol-1]))))) - (boundary value(S(soln) [non-dim]) * exp(-Symmetry factor [non-dim] * (Effective Voltage [non-dim] - (Reversible Potential [V] / (Gas constant [J K-1 mol-1] * Temperature [K] / Faraday Constant [C mol-1])))))) + Capacitance [F] * Electrode Area [cm2] * Gas constant [J K-1 mol-1] * Temperature [K] / Faraday Constant [C mol-1] / (Faraday Constant [C mol-1] * Electrode Area [cm2] * Far-field concentration of S(soln) [mol cm-3] * sqrt(Diffusion Coefficient of S [cm2 s-1]) * sqrt(Faraday Constant [C mol-1] * Scan Rate [V s-1] / (Gas constant [J K-1 mol-1] * Temperature [K])) * Gas constant [J K-1 mol-1] * Temperature [K] / Faraday Constant [C mol-1] / Scan Rate [V s-1]) * (((Voltage start [V] / (Gas constant [J K-1 mol-1] * Temperature [K] / Faraday Constant [C mol-1])) - (Voltage reverse [V] / (Gas constant [J K-1 mol-1] * Temperature [K] / Faraday Constant [C mol-1])) < time) - (time <= (Voltage start [V] / (Gas constant [J K-1 mol-1] * Temperature [K] / Faraday Constant [C mol-1])) - (Voltage reverse [V] / (Gas constant [J K-1 mol-1] * Temperature [K] / Faraday Constant [C mol-1]))))))', 'Effective Voltage [non-dim]'], domains={})

During handling of the above exception, another exception occurred:

Traceback (most recent call last):

  File ~\anaconda3\envs\pybamm\lib\site-packages\pybamm\parameters\parameter_values.py:497 in process_symbol
    return self._processed_symbols[symbol]

KeyError: Subtraction(-0x36494c20ced37b63, -, children=['Voltage start [V] / (Gas constant [J K-1 mol-1] * Temperature [K] / Faraday Constant [C mol-1]) + (time <= (Voltage start [V] / (Gas constant [J K-1 mol-1] * Temperature [K] / Faraday Constant [C mol-1])) - (Voltage reverse [V] / (Gas constant [J K-1 mol-1] * Temperature [K] / Faraday Constant [C mol-1]))) * -time + ((Voltage start [V] / (Gas constant [J K-1 mol-1] * Temperature [K] / Faraday Constant [C mol-1])) - (Voltage reverse [V] / (Gas constant [J K-1 mol-1] * Temperature [K] / Faraday Constant [C mol-1])) < time) * (time - (2.0 * ((Voltage start [V] / (Gas constant [J K-1 mol-1] * Temperature [K] / Faraday Constant [C mol-1])) - (Voltage reverse [V] / (Gas constant [J K-1 mol-1] * Temperature [K] / Faraday Constant [C mol-1])))))', 'Uncompensated Resistance [Ohm] * Faraday Constant [C mol-1] * Electrode Area [cm2] * Far-field concentration of S(soln) [mol cm-3] * sqrt(Diffusion Coefficient of S [cm2 s-1]) * sqrt(Faraday Constant [C mol-1] * Scan Rate [V s-1] / (Gas constant [J K-1 mol-1] * Temperature [K])) / (Gas constant [J K-1 mol-1] * Temperature [K] / Faraday Constant [C mol-1]) * (Redox Rate [s-1] * sqrt((Gas constant [J K-1 mol-1] * Temperature [K] * Diffusion Coefficient of S [cm2 s-1] / Faraday Constant [C mol-1]) * Scan Rate [V s-1]) / Diffusion Coefficient of S [cm2 s-1] * ((boundary value(P(soln) [non-dim]) * exp((1.0 - Symmetry factor [non-dim]) * (Effective Voltage [non-dim] - (Reversible Potential [V] / (Gas constant [J K-1 mol-1] * Temperature [K] / Faraday Constant [C mol-1]))))) - (boundary value(S(soln) [non-dim]) * exp(-Symmetry factor [non-dim] * (Effective Voltage [non-dim] - (Reversible Potential [V] / (Gas constant [J K-1 mol-1] * Temperature [K] / Faraday Constant [C mol-1])))))) + Capacitance [F] * Electrode Area [cm2] * Gas constant [J K-1 mol-1] * Temperature [K] / Faraday Constant [C mol-1] / (Faraday Constant [C mol-1] * Electrode Area [cm2] * Far-field concentration of S(soln) [mol cm-3] * sqrt(Diffusion Coefficient of S [cm2 s-1]) * sqrt(Faraday Constant [C mol-1] * Scan Rate [V s-1] / (Gas constant [J K-1 mol-1] * Temperature [K])) * Gas constant [J K-1 mol-1] * Temperature [K] / Faraday Constant [C mol-1] / Scan Rate [V s-1]) * (((Voltage start [V] / (Gas constant [J K-1 mol-1] * Temperature [K] / Faraday Constant [C mol-1])) - (Voltage reverse [V] / (Gas constant [J K-1 mol-1] * Temperature [K] / Faraday Constant [C mol-1])) < time) - (time <= (Voltage start [V] / (Gas constant [J K-1 mol-1] * Temperature [K] / Faraday Constant [C mol-1])) - (Voltage reverse [V] / (Gas constant [J K-1 mol-1] * Temperature [K] / Faraday Constant [C mol-1])))))'], domains={})

During handling of the above exception, another exception occurred:

Traceback (most recent call last):

  File ~\anaconda3\envs\pybamm\lib\site-packages\pybamm\parameters\parameter_values.py:497 in process_symbol
    return self._processed_symbols[symbol]

KeyError: Multiplication(0x28f5f51cf9799297, *, children=['Uncompensated Resistance [Ohm] * Faraday Constant [C mol-1] * Electrode Area [cm2] * Far-field concentration of S(soln) [mol cm-3] * sqrt(Diffusion Coefficient of S [cm2 s-1]) * sqrt(Faraday Constant [C mol-1] * Scan Rate [V s-1] / (Gas constant [J K-1 mol-1] * Temperature [K])) / (Gas constant [J K-1 mol-1] * Temperature [K] / Faraday Constant [C mol-1])', 'Redox Rate [s-1] * sqrt((Gas constant [J K-1 mol-1] * Temperature [K] * Diffusion Coefficient of S [cm2 s-1] / Faraday Constant [C mol-1]) * Scan Rate [V s-1]) / Diffusion Coefficient of S [cm2 s-1] * ((boundary value(P(soln) [non-dim]) * exp((1.0 - Symmetry factor [non-dim]) * (Effective Voltage [non-dim] - (Reversible Potential [V] / (Gas constant [J K-1 mol-1] * Temperature [K] / Faraday Constant [C mol-1]))))) - (boundary value(S(soln) [non-dim]) * exp(-Symmetry factor [non-dim] * (Effective Voltage [non-dim] - (Reversible Potential [V] / (Gas constant [J K-1 mol-1] * Temperature [K] / Faraday Constant [C mol-1])))))) + Capacitance [F] * Electrode Area [cm2] * Gas constant [J K-1 mol-1] * Temperature [K] / Faraday Constant [C mol-1] / (Faraday Constant [C mol-1] * Electrode Area [cm2] * Far-field concentration of S(soln) [mol cm-3] * sqrt(Diffusion Coefficient of S [cm2 s-1]) * sqrt(Faraday Constant [C mol-1] * Scan Rate [V s-1] / (Gas constant [J K-1 mol-1] * Temperature [K])) * Gas constant [J K-1 mol-1] * Temperature [K] / Faraday Constant [C mol-1] / Scan Rate [V s-1]) * (((Voltage start [V] / (Gas constant [J K-1 mol-1] * Temperature [K] / Faraday Constant [C mol-1])) - (Voltage reverse [V] / (Gas constant [J K-1 mol-1] * Temperature [K] / Faraday Constant [C mol-1])) < time) - (time <= (Voltage start [V] / (Gas constant [J K-1 mol-1] * Temperature [K] / Faraday Constant [C mol-1])) - (Voltage reverse [V] / (Gas constant [J K-1 mol-1] * Temperature [K] / Faraday Constant [C mol-1]))))'], domains={})

During handling of the above exception, another exception occurred:

Traceback (most recent call last):

  File ~\anaconda3\envs\pybamm\lib\site-packages\pybamm\parameters\parameter_values.py:497 in process_symbol
    return self._processed_symbols[symbol]

KeyError: Addition(-0x68cfe33ab0ae7fee, +, children=['Redox Rate [s-1] * sqrt((Gas constant [J K-1 mol-1] * Temperature [K] * Diffusion Coefficient of S [cm2 s-1] / Faraday Constant [C mol-1]) * Scan Rate [V s-1]) / Diffusion Coefficient of S [cm2 s-1] * ((boundary value(P(soln) [non-dim]) * exp((1.0 - Symmetry factor [non-dim]) * (Effective Voltage [non-dim] - (Reversible Potential [V] / (Gas constant [J K-1 mol-1] * Temperature [K] / Faraday Constant [C mol-1]))))) - (boundary value(S(soln) [non-dim]) * exp(-Symmetry factor [non-dim] * (Effective Voltage [non-dim] - (Reversible Potential [V] / (Gas constant [J K-1 mol-1] * Temperature [K] / Faraday Constant [C mol-1]))))))', 'Capacitance [F] * Electrode Area [cm2] * Gas constant [J K-1 mol-1] * Temperature [K] / Faraday Constant [C mol-1] / (Faraday Constant [C mol-1] * Electrode Area [cm2] * Far-field concentration of S(soln) [mol cm-3] * sqrt(Diffusion Coefficient of S [cm2 s-1]) * sqrt(Faraday Constant [C mol-1] * Scan Rate [V s-1] / (Gas constant [J K-1 mol-1] * Temperature [K])) * Gas constant [J K-1 mol-1] * Temperature [K] / Faraday Constant [C mol-1] / Scan Rate [V s-1]) * (((Voltage start [V] / (Gas constant [J K-1 mol-1] * Temperature [K] / Faraday Constant [C mol-1])) - (Voltage reverse [V] / (Gas constant [J K-1 mol-1] * Temperature [K] / Faraday Constant [C mol-1])) < time) - (time <= (Voltage start [V] / (Gas constant [J K-1 mol-1] * Temperature [K] / Faraday Constant [C mol-1])) - (Voltage reverse [V] / (Gas constant [J K-1 mol-1] * Temperature [K] / Faraday Constant [C mol-1]))))'], domains={})

During handling of the above exception, another exception occurred:

Traceback (most recent call last):

  File ~\anaconda3\envs\pybamm\lib\site-packages\pybamm\parameters\parameter_values.py:497 in process_symbol
    return self._processed_symbols[symbol]

KeyError: Multiplication(0x1a37bf8b0c0a9d96, *, children=['Capacitance [F] * Electrode Area [cm2] * Gas constant [J K-1 mol-1] * Temperature [K] / Faraday Constant [C mol-1] / (Faraday Constant [C mol-1] * Electrode Area [cm2] * Far-field concentration of S(soln) [mol cm-3] * sqrt(Diffusion Coefficient of S [cm2 s-1]) * sqrt(Faraday Constant [C mol-1] * Scan Rate [V s-1] / (Gas constant [J K-1 mol-1] * Temperature [K])) * Gas constant [J K-1 mol-1] * Temperature [K] / Faraday Constant [C mol-1] / Scan Rate [V s-1])', '((Voltage start [V] / (Gas constant [J K-1 mol-1] * Temperature [K] / Faraday Constant [C mol-1])) - (Voltage reverse [V] / (Gas constant [J K-1 mol-1] * Temperature [K] / Faraday Constant [C mol-1])) < time) - (time <= (Voltage start [V] / (Gas constant [J K-1 mol-1] * Temperature [K] / Faraday Constant [C mol-1])) - (Voltage reverse [V] / (Gas constant [J K-1 mol-1] * Temperature [K] / Faraday Constant [C mol-1])))'], domains={})

During handling of the above exception, another exception occurred:

Traceback (most recent call last):

  File ~\anaconda3\envs\pybamm\lib\site-packages\pybamm\expression_tree\symbol.py:758 in evaluate_for_shape
    return self._saved_evaluate_for_shape

AttributeError: 'Subtraction' object has no attribute '_saved_evaluate_for_shape'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):

  File ~\anaconda3\envs\pybamm\lib\site-packages\spyder_kernels\py3compat.py:356 in compat_exec
    exec(code, globals, locals)

  File c:\users\natha\desktop\code\main.py:143
    main()

  File c:\users\natha\desktop\code\main.py:96 in main
    cmodel = cm.CatalyticModel(const_parameters,seioptions, atol, rtol, t_steps, x_steps)

  File ~\Desktop\Code\catalyticmodel05.py:182 in __init__
    param.process_model(model)

  File ~\anaconda3\envs\pybamm\lib\site-packages\pybamm\parameters\parameter_values.py:341 in process_model
    new_algebraic[new_variable] = self.process_symbol(equation)

  File ~\anaconda3\envs\pybamm\lib\site-packages\pybamm\parameters\parameter_values.py:499 in process_symbol
    processed_symbol = self._process_symbol(symbol)

  File ~\anaconda3\envs\pybamm\lib\site-packages\pybamm\parameters\parameter_values.py:610 in _process_symbol
    new_left = self.process_symbol(symbol.left)

  File ~\anaconda3\envs\pybamm\lib\site-packages\pybamm\parameters\parameter_values.py:499 in process_symbol
    processed_symbol = self._process_symbol(symbol)

  File ~\anaconda3\envs\pybamm\lib\site-packages\pybamm\parameters\parameter_values.py:611 in _process_symbol
    new_right = self.process_symbol(symbol.right)

  File ~\anaconda3\envs\pybamm\lib\site-packages\pybamm\parameters\parameter_values.py:499 in process_symbol
    processed_symbol = self._process_symbol(symbol)

  File ~\anaconda3\envs\pybamm\lib\site-packages\pybamm\parameters\parameter_values.py:611 in _process_symbol
    new_right = self.process_symbol(symbol.right)

  File ~\anaconda3\envs\pybamm\lib\site-packages\pybamm\parameters\parameter_values.py:499 in process_symbol
    processed_symbol = self._process_symbol(symbol)

  File ~\anaconda3\envs\pybamm\lib\site-packages\pybamm\parameters\parameter_values.py:611 in _process_symbol
    new_right = self.process_symbol(symbol.right)

  File ~\anaconda3\envs\pybamm\lib\site-packages\pybamm\parameters\parameter_values.py:499 in process_symbol
    processed_symbol = self._process_symbol(symbol)

  File ~\anaconda3\envs\pybamm\lib\site-packages\pybamm\parameters\parameter_values.py:613 in _process_symbol
    new_symbol = symbol._binary_new_copy(new_left, new_right)

  File ~\anaconda3\envs\pybamm\lib\site-packages\pybamm\expression_tree\binary_operators.py:102 in _binary_new_copy
    return self._binary_evaluate(left, right)

  File ~\anaconda3\envs\pybamm\lib\site-packages\pybamm\expression_tree\binary_operators.py:273 in _binary_evaluate
    return left * right

  File ~\anaconda3\envs\pybamm\lib\site-packages\pybamm\expression_tree\symbol.py:559 in __mul__
    return pybamm.simplified_multiplication(self, other)

  File ~\anaconda3\envs\pybamm\lib\site-packages\pybamm\expression_tree\binary_operators.py:967 in simplified_multiplication
    if left.shape_for_testing == right.shape_for_testing and all(

  File ~\anaconda3\envs\pybamm\lib\site-packages\pybamm\expression_tree\symbol.py:942 in shape_for_testing
    evaluated_self = self.evaluate_for_shape()

  File ~\anaconda3\envs\pybamm\lib\site-packages\pybamm\expression_tree\symbol.py:760 in evaluate_for_shape
    self._saved_evaluate_for_shape = self._evaluate_for_shape()

  File ~\anaconda3\envs\pybamm\lib\site-packages\pybamm\expression_tree\binary_operators.py:114 in _evaluate_for_shape
    return self._binary_evaluate(left, right)

  File ~\anaconda3\envs\pybamm\lib\site-packages\pybamm\expression_tree\binary_operators.py:234 in _binary_evaluate
    return left - right

TypeError: numpy boolean subtract, the `-` operator, is not supported, use the bitwise_xor, the `^` operator, or the logical_xor function instead.

**Comment by brosaplanella** [comment]
The error seems to be from PyBaMM only, so the other packages and the environment should hopefully not affect it. From what we discussed in the other issue, the problem seemed to be with `E_app`, but a minimum working example (as a single script, with the minimum code to reproduce the error) would help).

**Reply by nril0001** [reply]
So in true coding fashion after I posted this I worked out what the issue is, just not how to fix it. The issue seems to be coming from the Eapp term. I am performing a voltage sweep which requires the Eapp code below and in order to model current correctly I need to be able to acquire the time derivative of Eapp, I've discovered that attempting to do so with

Eapp.diff(pybamm.t) produces the error above. 

I've zipped a basic example below

# Input voltage protocol
        Edc_forward = -pybamm.t
        Edc_backwards = pybamm.t - 2 * t_reverse
        Eapp = E_start + \
            (pybamm.t <= t_reverse) * Edc_forward + \
            (pybamm.t > t_reverse) * Edc_backwards      

i = Cdl * Eapp.diff(pybamm.t)
[Code.zip](https://github.com/pybamm-team/PyBaMM/files/11944914/Code.zip)

**Reply by nril0001** [reply]
something to note is this code works in an ubuntu environment that is running pybamm 0.2.3 ( I don't know why the version is so low, it is just the one that installed for me on that environment)

**Reply by brosaplanella** [reply]
Which Python version are you using?

**Reply by nril0001** [reply]
Both are running on 3.10.11

These are the backup files for both environments that anaconda produces they list the version of each package
[Code.zip](https://github.com/pybamm-team/PyBaMM/files/11956160/Code.zip)

and I'm not sure if this is related but this is the message that I keep getting if I attempt to use conda to install pybamm.

In particular it is the final specification that is consistent in both Ubuntu and Windows, despite the fact that I have sympy installed in the environment and I have made sure of it, it still doesn't seem to register. This may be an unrelated problem however

![image](https://github.com/pybamm-team/PyBaMM/assets/89230336/7e2fdcb3-c695-4def-91d0-987a027997b6)

**Reply by brosaplanella** [reply]
Python 3.10 is only supported from the most recent version (23.5), so I would suggest to upgrade your version and see if the errors persist.

**Reply by nril0001** [reply]
My pybamm on the windows environment is already upgraded to the most recent version. I get the error on pybamm 23.5 with python 3.10. I have also tried previous versions of pybamm such as 23.4, 23.3 and 23.2 on 3.9 or 3.8 and I was still getting the error

**Reply by brosaplanella** [reply]
The main issue is that you should reformat your model as an ODE for the voltage, as @tinosulzer suggested in the other discussion: https://github.com/pybamm-team/PyBaMM/discussions/3027#discussioncomment-6180060.

Switching to that formulation is a necessary condition to get the code working, even if new errors arise.
